### DISCUSSION TITLE
Chart subplots with bidirectional binding of xaxis

### DISCUSSION THREAD

**Discussion by raven2cz** [GSI-CS-CO/chart-fx#392] [Ideas]
I forgot to check next important feature. It seems that you don't need it or I miss it in the samples. It's next very important feature for chart library. 
It's called **subplots**. The final plot is assembly from more subplots which have same x-axis with bidirectional behavior for all subplots included. If the zoomer / slider is moved all subplots are still part of same plot and it is repainted. 
I found sample _DimReductionDataSetSample_ where is the second column plots bounded together. But it is still two independent plots. I tried to hide top x-axis with xaxis.setVisible(false), but nothing happen, the axis is still visible. 
Do you plan this? Or can you create some simple sample where are main plot and second smaller plot under it with same x-axis under it to handle whole chart?
There is simple example from my previous jfreechart library.

![subplots](https://user-images.githubusercontent.com/20504243/100422612-ee813f80-308a-11eb-9046-c1e0966bf7ab.png)

**Comment by wirew0rm** [comment]
Hey,

yes, you are right, we do not really have extensive support for subplots.

Here is what is currently available and what we use in our applications:
- `de.gsi.chart.utils.AxisSynchronizer`bidirectional synchronisation of the axis range between two plots, as used in the sample you mentioned.
- Currently it is not possible to hide the axis, so you would have to set empty labels and invisible tick marks. The axes needs to be there because the coordinates are computed based on the layout of the axis. I already thought about implementing a blank axis without any marks and labels for such usecases, which should be pretty straightforward.
- For stacking charts there is the problem of different axis width for the y axis, so you would have to make sure to set a sufficient and equal width to all of them (or do some complex bindings without creating an infinite loop).

To help with these layout questions we have the ChartAnatomySample, to show the different components:
![chartAnatomy](https://raw.githubusercontent.com/GSI-CS-CO/chart-fx/master/docs/pics/ChartAnatomySample.png)

Correctly implementing subplots would be a major undertaking and would probably break existing apps. Additionally we would have to think about how plugins (zoomer, tooltip, parameter measurements, indicators) should handle this, the mechanics for updating axes shared between different (sub)charts, etc. (I recently extended the Indicators to correctly handle multiple axes in a single chart, even that case is not so simple.) It is definitely interesting and useful functionality, but its not high enough priority for us at the moment compared to the amount of work it would need to properly implement.

I'll leave this feature request open to collect workarounds and implementation details for a potential future implementation.

**Comment by raven2cz** [comment]
Thanks Alexander,
I will try tests, and maybe it is not problem and AxisSynchronizer + hidden empty labels/invisible ticks will be accurate. 
In any cases, I have to create some separate "better" slider which will be placed under all plots. I found the problem that actual slider has very "very very" small for longer history, so it cannot be move, because second slider feature for extending range is activate instead of movement feature.

**Comment by TA-code1973** [comment]
> Thanks Alexander,
> I will try tests, and maybe it is not problem and AxisSynchronizer + hidden empty labels/invisible ticks will be accurate.
> In any cases, I have to create some separate "better" slider which will be placed under all plots. I found the problem that actual slider has very "very very" small for longer history, so it cannot be move, because second slider feature for extending range is activate instead of movement feature.

Hi Raven,

Maybe you can take a look a the binjr timeseries browser. This is a javafx application with charts and corresponding subplots.
Link is : https://github.com/binjr/binjr
But probably they don't use a custom renderer. But overall the quality of the java code is not bad.

Regards,
Arjen
